### DISCUSSION TITLE
Problem building for a specific environment

### DISCUSSION THREAD

**Discussion by phermsdorf** [eclipse-tycho/tycho#2601] [Q&A]
Hi,

in my parent pom's pluginManagement section i have a target-platform-configuration like this:
```
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>target-platform-configuration</artifactId>
    <version>${tycho-version}</version>
    <configuration>
            <executionEnvironment>JavaSE-17</executionEnvironment>
            <environments>
                <environment>
                    <os>linux</os>
                    <ws>gtk</ws>
                    <arch>x86_64</arch>
                </environment>
                <environment>
                    <os>win32</os>
                    <ws>win32</ws>
                    <arch>x86_64</arch>
                </environment>
            </environments>
            <dependency-resolution>
                <profileProperties>
                <org.eclipse.justj.buildtime>true</org.eclipse.justj.buildtime>
                </profileProperties>
            </dependency-resolution>
    </configuration>
</plugin>
```
in a profile i want to be able to just build for one environment (win32 in this case) and try to override that "global" configuration with:
```
<profile>
    <id>target-rcp-win32</id>
    <build>
        <plugins>
            <plugin>
                <groupId>org.eclipse.tycho</groupId>
                <artifactId>target-platform-configuration</artifactId>
                <version>${tycho-version}</version>
                <configuration>
                        <environments>
                            <environment>
                                <os>win32</os>
                                <ws>win32</ws>
                                <arch>x86_64</arch>
                            </environment>
                        </environments>
                </configuration>
            </plugin>
        </plugins>
    </build>
</profile>
```
when building with this profile enabled i get the following error message:
`Failed to execute goal org.eclipse.tycho:tycho-packaging-plugin:3.0.4:package-feature(default-package-feature) on project com.godyo.p4.runtime.feature: Execution default-package-feature of goal org.eclipse.tycho:tycho-packaging-plugin:3.0.4:package-feature failed: eclipse-plugin artifact with ID "com.godyo.p5.ui.gtk.linux.x86_64" and version matching "0.0.0" was not found in the target platform`

in the feature the plugin is defined with platform filters:
```
   <plugin
         id="com.godyo.p5.ui.gtk.linux.x86_64"
         os="linux"
         ws="gtk"
         arch="x86_64"
         download-size="0"
         install-size="0"
         version="0.0.0"
         unpack="false"/>
```
The plugin itself also has a platform filter in the MANIFEST.MF like this: `(& (osgi.ws=gtk) (osgi.os=linux) (osgi.arch=x86_64))`

Shouldn't this configuration work and just skip this plugin since it does not match the selected environment? 

Thanks for any hints on this.
Bye Peter

**Comment by laeubi** [comment]
You might want to let maven show you the effective pom, I think in this case you need to override the configuration from the parent:

https://blog.sonatype.com/2011/01/maven-how-to-merging-plugin-configuration-in-complex-projects/

**Reply by phermsdorf** [reply]
Thanks for the pointer - that was new to me.

I added to the definition inside the profile like this:
```
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>target-platform-configuration</artifactId>
    <configuration>
            <environments combine.self="override">
                <environment>
                    <os>win32</os>
                    <ws>win32</ws>
                    <arch>x86_64</arch>
                </environment>
            </environments>
    </configuration>
</plugin>
```
but the problem persists.
The effective pom for the feature show's this setup for the target-platform-configuration:
```
<pluginManagement>
  <!-- identical to parent pom with both environments -->
</pluginManagement>
<plugins>
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>target-platform-configuration</artifactId>
    <version>3.0.4</version>
    <executions>
    <execution>
        <id>default-target-platform</id>
        <phase>initialize</phase>
        <goals>
        <goal>target-platform</goal>
        </goals>
        <configuration>
        <environments combine.self="override">
            <environment>
            <os>win32</os>
            <ws>win32</ws>
            <arch>x86_64</arch>
            </environment>
        </environments>
        <target>
            <artifact>
            <groupId>com.godyo.p4</groupId>
            <artifactId>com.godyo.p5.target</artifactId>
            <version>4.1.0-SNAPSHOT</version>
            </artifact>
        </target>
        <executionEnvironment>JavaSE-17</executionEnvironment>
        </configuration>
    </execution>
    </executions>
    <configuration>
    <environments combine.self="override">
        <environment>
        <os>win32</os>
        <ws>win32</ws>
        <arch>x86_64</arch>
        </environment>
    </environments>
    <target>
        <artifact>
        <groupId>com.godyo.p4</groupId>
        <artifactId>com.godyo.p5.target</artifactId>
        <version>4.1.0-SNAPSHOT</version>
        </artifact>
    </target>
    <executionEnvironment>JavaSE-17</executionEnvironment>
    </configuration>
</plugin>
</plugins>
```
like expected with only one environment configuration now.
The effective pom of the "missing" plugin show the following target-platform.configuration:
```
<plugin>
    <groupId>org.eclipse.tycho</groupId>
    <artifactId>target-platform-configuration</artifactId>
    <version>3.0.4</version>
    <executions>
    <execution>
        <id>default-target-platform</id>
        <phase>initialize</phase>
        <goals>
        <goal>target-platform</goal>
        </goals>
        <configuration>
        <environments combine.self="override">
            <environment>
            <os>linux</os>
            <ws>gtk</ws>
            <arch>x86_64</arch>
            </environment>
        </environments>
        <target>
            <artifact>
            <groupId>com.godyo.p4</groupId>
            <artifactId>com.godyo.p5.target</artifactId>
            <version>4.1.0-SNAPSHOT</version>
            </artifact>
        </target>
        <executionEnvironment>JavaSE-17</executionEnvironment>
        </configuration>
    </execution>
    </executions>
    <configuration>
    <environments combine.self="override">
        <environment>
        <os>linux</os>
        <ws>gtk</ws>
        <arch>x86_64</arch>
        </environment>
    </environments>
    <target>
        <artifact>
        <groupId>com.godyo.p4</groupId>
        <artifactId>com.godyo.p5.target</artifactId>
        <version>4.1.0-SNAPSHOT</version>
        </artifact>
    </target>
    
    <executionEnvironment>JavaSE-17</executionEnvironment>
    </configuration>
</plugin>
```
Does the linux execution environment comes from the platform filter definition inside the MANIFEST.MF?
Maybe it's thatswhy filtered out from the build reactor and later in the build not found ... even if it shouldn't be included in this environment anyway. But i'm just guessing.

I'm really wondering what's wrong here .... shouldn't it be daily business to leave out artifacts that do not belong to the current environment.....

**Reply by laeubi** [reply]
> I'm really wondering what's wrong here .... shouldn't it be daily business to leave out artifacts that do not belong to the current environment.....

Usually one simply builds all plugins together, Tyho can cross-compile plugins for different environments, neverless you maybe want to create a minimal example or provide an [integration-test](https://github.com/eclipse/tycho/wiki#providing-an-integration-test) to demonstrate the issue.

**Reply by phermsdorf** [reply]
> Usually one simply builds all plugins together, Tyho can cross-compile plugins for different environments

Sure. I've come this way because of a similar issue like [this](https://github.com/codehaus-plexus/plexus-archiver/issues/66#issue-240423954) while upgrading our product to the JustJ JRE Version 17 and building under windows (because of runnig swtbot tests there)

I'll see if i can produce a litte example project or modify an existing one
